kmultiversx/src/kmultiversx/kasmer.py

```diff
@@ -151,7 +164,34 @@ def run_config_and_check_empty(
     return final_conf, sym_conf, subst
 
 
-def run_test(krun: KRun, sym_conf: KInner, init_subst: dict[str, KInner], endpoint: str, args: tuple[str, ...]) -> None:
+def run_pattern_and_check_exit_code(krun: KRun, conf: Pattern, pipe_stderr: bool = False) -> None:
+    pattern_out = llvm_interpret_raw(krun.definition_dir, conf.text, pipe_stderr)
+
+    final_conf = parse_pattern(pattern_out)
+
+    exit_code_cell = final_conf.arguments[0].arguments[0].arguments[5]
+    assert exit_code_cell.constructor.name == "Lbl'-LT-'exit-code'-GT-'"
+    exit_code = int(exit_code_cell.arguments[0].arguments[0].contents)
+
+    if exit_code != 0:
+        kast_conf = krun.kore_to_kast(KoreParser(pattern_out).pattern())
+        sym_conf, subst = split_config_from(kast_conf)
+        k_cell = subst['K_CELL']
+        print(krun.pretty_print(k_cell))
+        raise KasmerRunError(
+            k_cell=subst['K_CELL'],
+            vm_output=subst['VMOUTPUT_CELL'],
+            logging=subst['LOGGING_CELL'],
+            final_conf=kast_conf,
+            message='k cell not empty',
+        )
```

Comment:
If the test fails, these lines are never executed. `llvm_interpret_raw` calls `run_process` with `check=True` by default, so if the `<exit-code>` cell is non-zero it throws an exception because the cell has the `exit` attribute.
```
<exit-code exit=""> 0 </exit-code>
``` 

I'll send my code suggestions in a separate comment.

Reply:
Nice catch, thank you!

Performance didn't improve much, because the bindings parser was already very fast, but every little bit helps.